hilti/toolchain/src/compiler/optimizer.cc

```diff
@@ -1557,9 +1593,105 @@ struct MemberVisitor : OptimizerVisitor {
     }
 };
 
+struct FunctionBodyVisitor : OptimizerVisitor {
+    using OptimizerVisitor::OptimizerVisitor;
+
+    bool prune_uses(Node* node) override {
+        visitor::visit(*this, node);
+        return isModified();
+    }
+
+    bool remove_node(detail::cfg::CFG& cfg, Node* data, const std::string& msg = {}) {
+        assert(data);
+
+        Node* dead = nullptr;
+
+        if ( data->isA<Statement>() && data->hasParent() )
+            dead = data;
+
+        else if ( data->isA<Expression>() ) {
+            auto* p = data->parent();
+
+            while ( p && ! p->isA<Statement>() )
+                p = p->parent();
+
+            if ( p && p->hasParent() )
+                dead = p;
+        }
+
+        if ( dead ) {
+            // Edit AST.
+            removeNode(dead, msg);
+
+            // Make equivalent edit to control flow graph.
+            const auto& out = cfg.g.neighborsDownstream(data->identity());
+            const auto& in = cfg.g.neighborsUpstream(data->identity());
+
+            // Create new edges between incoming and outgoing nodes.
+            for ( auto&& i : in ) {
+                for ( auto&& o : out )
+                    cfg.g.addEdge(i, o);
+            }
+
+            cfg.g.removeNode(data->identity());
+
+            return true;
+        }
+
+        return false;
+    }
+
+    void visit_node(Node* n) {
+        while ( true ) {
+            bool modified = false;
+
+            // FIXME(bbannier): In principal we should be able to reuse the
+            // flow through optimizations, but this currently fails due to
+            // edits not correctly changing the flow.
+            auto cfg = detail::cfg::CFG(n);
+            cfg.populateReachableExpressions();
+
+            for ( auto&& x : cfg.unreachableStatements() )
+                modified |= remove_node(cfg, x, "statement result unused");
+
+            if ( modified )
+                break;
+
+            auto unreachable_nodes = cfg.unreachableNodes();
+            if ( unreachable_nodes.empty() )
+                break;
+
+            // Remove unreachable control flow branches.
+            // NOLINTNEXTLINE(bugprone-nondeterministic-pointer-iteration-order)
+            for ( auto&& n : unreachable_nodes )
+                modified |= remove_node(cfg, n, "unreachable code");
+
+            if ( ! modified )
+                break;
+        }
+    }
+
+    void operator()(declaration::Function* f) override {
+        if ( auto&& body = f->function()->body() )
+            visit_node(body);
```

Comment:
I think this covers everything but inline methods, this doesn't end up getting optimizations:

```
type S = struct {
    method void f() { } # No CFG!
};
```